pypesto/ensemble/utils.py

```diff
@@ -0,0 +1,187 @@
+                        ensemble_type=ensemble_type,
+                        lower_bound=lower_bound,
+                        upper_bound=upper_bound)
+
+
+def read_from_df(dataframe: pd.DataFrame,
+                 headline_parser: Callable = None,
+                 ensemble_type: EnsembleType = None,
+                 lower_bound: np.ndarray = None,
+                 upper_bound: np.ndarray = None):
+    """
+    function for creating an ensemble from a csv file
+
+    Parameters
+    ----------
+    dataframe:
+        pandas.DataFrame to read in parameter ensemble
+    headline_parser:
+        A function which reads in the headline of the csv file and converts it
+        into vector_tags (see constructor of Ensemble for more details)
+    ensemble_type:
+        Ensemble type: representative sample or random ensemble
+    lower_bound:
+        array of potential lower bounds for the parameters
+    upper_bound:
+        array of potential upper bounds for the parameters
+
+    Returns
+    -------
+    result:
+        Ensemble object of parameter vectors
+    """
+    # if we have a parser to make vector_tags from column names, we use it
+    vector_tags = None
+    if headline_parser is not None:
+        vector_tags = headline_parser(list(dataframe.columns))
+    # set the type of the ensemble
+    if ensemble_type is None:
+        ensemble_type = EnsembleType.ensemble
+
+    return Ensemble(x_vectors=dataframe.values,
+                    x_names=list(dataframe.index),
+                    vector_tags=vector_tags,
+                    ensemble_type=ensemble_type,
+                    lower_bound=lower_bound,
+                    upper_bound=upper_bound)
+
+
+def write_ensemble_prediction_to_h5(ensemble_prediction: EnsemblePrediction,
+                                    output_file: str,
+                                    base_path: str = None):
+    # parse base path
+    if base_path is not None:
+        base = base_path
+        if base_path[-1] != '/':
+            base = base_path + '/'
+    else:
+        base = ''
+
+    # open file
+    f = h5py.File(output_file, 'w')
+
+    # write prediction ID if available
+    if ensemble_prediction.prediction_id is not None:
+        f.create_dataset(base + PREDICTION_ID,
+                         data=ensemble_prediction.prediction_id)
+
+    # write the single prediction results
+    for i_result, result in enumerate(ensemble_prediction.prediction_results):
+        tmp_base_path = base + f'{PREDICTION_RESULTS}_{i_result}'
+        result.write_to_h5(output_file, base_path=tmp_base_path)
+
+    # write lower bounds per condition, if available
+    if ensemble_prediction.lower_bound is not None:
+        f.create_group(base + f'{LOWER_BOUND}s')
+        for i_cond, lower_bounds in enumerate(ensemble_prediction.lower_bound):
+            condition_id = \
+                ensemble_prediction.prediction_results[0].condition_ids[i_cond]
```

Comment:
In principle yes, but we store a multitude of PredictionResults here. As PredictionResult also must work as a class of its own, it makes sense to also store the conditionIds in there. One could now either reduce what's stored, therefore introduce a new class... or storing something redundant... What do you think?

Reply:
For the purposes of e.g. plotting by condition or observable, it would be nice to automatically ensure that all `PredictionResults` share the same condition IDs, and the prediction for the same condition between different `PredictionResults` contains the same observable IDs. And also to provide methods to easily get all data by condition or observable, from an `EnsemblePrediction`. Maybe a wrapper for `SampleEnsemblePrediction`? Anyway, I can do this, let me know what you think

Reply:
I would actually suggest to open 3 issues:
 - wishes and stuff we can use the ensemble functionality for (e.g., what you just wrote)
 - wishes and stuff we can use the prediction functionality for (e.g., prediction profiles)
 - wishes and stuff we can use the dimesnion reduction functionality for (e.g., analyzing optimization runs from a multi-start visually)

Reply:
But yes: Asserting these things are equal would be good. Currently, they must be equal, just by the way the predict functions creates things...